backends/graphics/dispmanxsdl/dispmanxsdl-graphics.cpp

```diff
@@ -0,0 +1,589 @@
+	vc_dispmanx_rect_set(&_dispvars->srcRect, 0, 0, srcWidth << 16, srcHeight << 16);
+
+	for (int i = 0; i < _dispvars->numpages; i++) {
+		_dispvars->pages[i].resource = vc_dispmanx_resource_create(_dispvars->pixFormat,
+			srcWidth, srcHeight, &(_dispvars->vcImagePtr));
+	}
+
+	// Add the element. Has to be removed before getting here again.
+	_dispvars->update = vc_dispmanx_update_start(0);
+
+	_dispvars->element = vc_dispmanx_element_add(
+		_dispvars->update,_dispvars->display, 0,
+		&_dispvars->dstRect, 0,
+		&_dispvars->srcRect, DISPMANX_PROTECTION_NONE,
+		&_dispvars->alpha, 0, (DISPMANX_TRANSFORM_T)0);
+
+	vc_dispmanx_update_submit_sync(_dispvars->update);
+}
+
+void dispmanXVSyncCallback (DISPMANX_UPDATE_HANDLE_T u, void *arg) {
+	struct dispmanxPage *page = (struct dispmanxPage*)arg;
+	struct dispvarsStruct *dispvars = page->dispvars;
+
+	// Marking the page as free must be done before the signaling
+	// so when the update function continues (it won't continue until we signal)
+	// we can chose this page as free.
+	if (dispvars->currentPage) {
+		pthread_mutex_lock(&dispvars->currentPage->pageUsedMutex);
+
+		// We mark as free the page that was visible until now.
+		page->dispvars->currentPage->used = false;
+
+		pthread_mutex_unlock(&dispvars->currentPage->pageUsedMutex);
+	}
+
+	// The page on which we issued the flip that
+	// caused this callback becomes the visible one
+	dispvars->currentPage = page;
+
+	// These two things must be isolated "atomically" to avoid getting
+	// a false positive in the pending_mutex test in update function.
+	pthread_mutex_lock(&dispvars->pendingMutex);
+
+	dispvars->pageflipPending--;
+	pthread_cond_signal(&dispvars->vsyncCondition);
+
+	pthread_mutex_unlock(&dispvars->pendingMutex);
+}
+
+void DispmanXSdlGraphicsManager::dispmanXUpdate() {
+	// Wait until last issued flip completes to get a free page. Also,
+	// dispmanx doesn't support issuing more than one pageflip.
+	pthread_mutex_lock(&_dispvars->pendingMutex);
+
+	if (_dispvars->pageflipPending > 0) {
+		pthread_cond_wait(&_dispvars->vsyncCondition, &_dispvars->pendingMutex);
+	}
+
+	pthread_mutex_unlock(&_dispvars->pendingMutex);
+
+	struct dispmanxPage *page = dispmanXGetFreePage();
+
+	// Frame blitting
+	vc_dispmanx_resource_write_data(page->resource, _dispvars->pixFormat,
+		_dispvars->pitch, _dispvars->pixmem, &_dispvars->bmpRect);
+
+	// Issue a page flip at the next vblank interval (will be done at vsync anyway).
+	_dispvars->update = vc_dispmanx_update_start(0);
+
+	vc_dispmanx_element_change_source(_dispvars->update, _dispvars->element,
+		page->resource);
+	vc_dispmanx_update_submit(_dispvars->update, &dispmanXVSyncCallback, page);
+
+	pthread_mutex_lock(&_dispvars->pendingMutex);
+	_dispvars->pageflipPending++;
+	pthread_mutex_unlock(&_dispvars->pendingMutex);
+}
+
+struct dispmanxPage *DispmanXSdlGraphicsManager::dispmanXGetFreePage(void) {
+	struct dispmanxPage *page = NULL;
+
+	while (!page)
+	{
+		// Try to find a free page
```

Comment:
Minor: opening brace shouldn't be on separate line according to ScummVM CFC. This occurs multiple times later on, I'll not bother you with pointing out each single one :).